rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingPublish);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::PUBACK(ref v)
+            | MQTTOperation::PUBCOMP(ref v) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(v.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingPublish);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::SUBACK(ref suback) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(suback.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingSubscribe);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::UNSUBACK(ref unsuback) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(unsuback.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingUnsubscribe);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::UNASSIGNED => {
+                let mut tx = self.new_tx(msg, toclient);
+                tx.complete = true;
+                &mut MQTTState::set_event(&mut tx, MQTTEvent::UnassignedMsgtype);
+                self.transactions.push(tx);
+            },
+            _ => {
+                if !self.connected {
```

Comment:
Can we reach here ?

Reply:
Yes, with the `PING*` message types. I will add them explicitly so this is visible in the code itself.

Reply:
thanks